<center>
<h1 style= "color:#FFC300">
    Segmenting and Clustering Neighborhoods in Toronto 
</h1> 

<img src="toronto.png">

In [1]:
#Packages
!pip install geopy folium
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup 
import requests
from geopy.geocoders import Nominatim 
import requests 
from pandas.io.json import json_normalize 
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import folium 
print("All packages are imported!")

All packages are imported!


### Creating a Soup object

In [2]:
url="https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
data=requests.get(url).text
soup = BeautifulSoup(data,"html5lib")

### Creating the dataset 
**1-Create an empty list**

**2- Find table in soup object and pass Not Assigned cells, get PostalCode, Borough and Neighbourhood and append them into cell**

**3-Append cell into table_contents**

**4-Convert list to a dataframe**

**5-Replace some names in Borough column**

In [3]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

In [4]:
df

Borough  \
0                North York   
1                North York   
2          Downtown Toronto   
3                North York   
4              Queen's Park   
5                 Etobicoke   
6               Scarborough   
7                North York   
8                 East York   
9          Downtown Toronto   
10               North York   
11                Etobicoke   
12              Scarborough   
13               North York   
14                East York   
15         Downtown Toronto   
16                     York   
17                Etobicoke   
18              Scarborough   
19             East Toronto   
20         Downtown Toronto   
21                     York   
22              Scarborough   
23                East York   
24         Downtown Toronto   
25         Downtown Toronto   
26              Scarborough   
27               North York   
28               North York   
29                East York   
..                      ...   
73          Central Toronto   
74          Central Toronto   
75             West Toronto   
76              Mississauga   
77                Etobicoke   
78              Scarborough   
79          Central Toronto   
80         Downtown Toronto   
81             West Toronto   
82              Scarborough   
83          Central Toronto   
84         Downtown Toronto   
85              Scarborough   
86          Central Toronto   
87         Downtown Toronto   
88                Etobicoke   
89                Etobicoke   
90              Scarborough   
91         Downtown Toronto   
92   Downtown Toronto Stn A   
93                Etobicoke   
94      Etobicoke Northwest   
95              Scarborough   
96         Downtown Toronto   
97         Downtown Toronto   
98                Etobicoke   
99         Downtown Toronto   
100   East Toronto Business   
101               Etobicoke   
102               Etobicoke   

                                          Neighborhood PostalCode  
0                                            Parkwoods        M3A  
1                                     Victoria Village        M4A  
2                            Regent Park, Harbourfront        M5A  
3                     Lawrence Manor, Lawrence Heights        M6A  
4                        Ontario Provincial Government        M7A  
5                                     Islington Avenue        M9A  
6                                       Malvern, Rouge        M1B  
7                                      Don Mills North        M3B  
8                      Parkview Hill, Woodbine Gardens        M4B  
9                             Garden District, Ryerson        M5B  
10                                           Glencairn        M6B  
11   West Deane Park, Princess Gardens, Martin Grov...        M9B  
12              Rouge Hill, Port Union, Highland Creek        M1C  
13                                     Don Mills South        M3C  
14                                    Woodbine Heights        M4C  
15                                      St. James Town        M5C  
16                                  Humewood-Cedarvale        M6C  
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...        M9C  
18                   Guildwood, Morningside, West Hill        M1E  
19                                         The Beaches        M4E  
20                                         Berczy Park        M5E  
21                                 Caledonia-Fairbanks        M6E  
22                                              Woburn        M1G  
23                                             Leaside        M4G  
24                                  Central Bay Street        M5G  
25                                            Christie        M6G  
26                                           Cedarbrae        M1H  
27                                   Hillcrest Village        M2H  
28     Bathurst Manor, Wilson Heights, Downsview North        M3H  
29                                    Thorncliffe Park   

In [5]:
df.shape

(103, 3)

There are 103 rows and 3 columns in dataset.

### Adding Latitude and the Longitude coordinates of each neighborhood. 

In [7]:
geodata = pd.read_csv("Geospatial_Coordinates.csv")
geodata

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
..          ...        ...        ...
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

**Firstly, change the name of the column Postal Code to PostalCode in order to have same name for the common column between datasets**

**Secondly, using pandas merge function to add Latitude and Longitude to the main dataset df**

In [8]:
geodata=geodata.rename(columns={'Postal Code':'PostalCode'})
df= pd.merge(df,geodata,on='PostalCode')
df.head()

Borough                      Neighborhood PostalCode   Latitude  \
0        North York                         Parkwoods        M3A  43.753259   
1        North York                  Victoria Village        M4A  43.725882   
2  Downtown Toronto         Regent Park, Harbourfront        M5A  43.654260   
3        North York  Lawrence Manor, Lawrence Heights        M6A  43.718518   
4      Queen's Park     Ontario Provincial Government        M7A  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

**Checking if the merging is correct, it can be seen taht Latitude and Longitude of M5G Postal Code are same in both datasets**

In [9]:
print( df[df['PostalCode']=='M5G'])
print(geodata[geodata['PostalCode']=='M5G'])

             Borough        Neighborhood PostalCode   Latitude  Longitude
24  Downtown Toronto  Central Bay Street        M5G  43.657952 -79.387383
   PostalCode   Latitude  Longitude
57        M5G  43.657952 -79.387383


### Using Geopy Library to get latitude and longitude values of Toronto

In [10]:
address = 'Toronto, Ontario'
geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


# Map of Toronto

In [11]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Filtering for Borough includes Toronto

In [12]:
toronto_data =  df[df['Borough'].str.contains("Toronto")]
toronto_data.head()

Borough               Neighborhood PostalCode   Latitude  \
2   Downtown Toronto  Regent Park, Harbourfront        M5A  43.654260   
9   Downtown Toronto   Garden District, Ryerson        M5B  43.657162   
15  Downtown Toronto             St. James Town        M5C  43.651494   
19      East Toronto                The Beaches        M4E  43.676357   
20  Downtown Toronto                Berczy Park        M5E  43.644771   

    Longitude  
2  -79.360636  
9  -79.378937  
15 -79.375418  
19 -79.293031  
20 -79.373306

### Map of filtered data (Boroughs include Toronto--center)

In [13]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [14]:
CLIENT_ID = 'NY5GVWOPLWYA4EJNZLIKNYI14RV1SQHJMFZN2JDS5H0AYHP1' # your Foursquare ID
CLIENT_SECRET = 'S5SGMCXEI5GB10OXX50XHI04RKNQBPELSGXPKRHSBN4K1OGA' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

### Function that gets venues in Toronto within a radius of 500 meters

In [17]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [18]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Enclave of M5E
St. James Town, Cabbagetown
First Canadi

In [24]:
print(toronto_venues.shape)
toronto_venues.head()

(1594, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3          Impact Kitchen       43.656369       -79.356980   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3           Restaurant  
4                  Spa

In [38]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
Brockton, Parkdale Village, Exhibition Place                           22   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                     68   
Christie                                                               16   
Church and Wellesley                                                   78   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           16   
Enclave of M4L                                                         17   
Enclave of M5E                                                         98   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         20   
Kensington Market, Chinatown, Grange Park                              67   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               42   
Moore Park, Summerhill East                                             2   
North Toronto West                                                     19   
Parkdale, Roncesvalles                                                 14   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     36   
St. James Town                                                         86   
St. James Town, Cabbagetown                                            44   
Studio District                                                        36   
Summerhill West, Rathnelly, South Hill, Forest ...                     13   
The Annex, North Midtown, Yorkville                                    19   
The Beaches                                                             4   
The Danforth  East                                                      3   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         33   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             58   
Brockton, Parkdale Village, Exhibition Place                            22   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                      68   
Christie                                                                16   
Church and Wellesley                                                    78   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              34   
Davisville North                                                         9   
Dufferin, Dovercourt Village        

In [27]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 231 uniques categories.


## Analyzing Neighborhoods

Creating dummies to look at what is in these neighbourhoods

In [59]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 
# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0            0        0                   0             0               0   
1            0        0                   0             0               0   
2            0        0                   0             0               0   
3            0        0                   0             0               0   
4            0        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  ...  Theme Restaurant  Tibetan Restaurant  Toy / Game Store  \
0         0  ...                 0                   0                 0   
1         0  ...                 0                   0                 0   
2         0  ...                 0                   0                 0   
3         0  ...                 0                   0                 0   
4         0  ...                 0                   0                 0   

   Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0      0              0                              0                 0   
1      0              0                              0                 0   
2      0              0                              0                 0   
3      0              0                              0                 0   
4      0              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  
0                      0         0          0  
1                      0         0          0  
2                      0         0          0  
3                      0         0          0  
4                      0         0          0  

[5 rows x 231 columns]

In [43]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio   Airport  \
0                                         Berczy Park     0.000000  0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000  0.000000   
2   CN Tower, King and Spadina, Railway Lands, Har...     0.000000  0.066667   
3                                  Central Bay Street     0.014706  0.000000   
4                                            Christie     0.000000  0.000000   
5                                Church and Wellesley     0.025641  0.000000   
6                      Commerce Court, Victoria Hotel     0.010000  0.000000   
7                                          Davisville     0.000000  0.000000   
8                                    Davisville North     0.000000  0.000000   
9                        Dufferin, Dovercourt Village     0.000000  0.000000   
10                                     Enclave of M4L     0.058824  0.000000   
11                                     Enclave of M5E     0.010204  0.000000   
12             First Canadian Place, Underground city     0.000000  0.000000   
13                           Forest Hill North & West     0.000000  0.000000   
14                           Garden District, Ryerson     0.000000  0.000000   
15  Harbourfront East, Union Station, Toronto Islands     0.000000  0.000000   
16                      High Park, The Junction South     0.000000  0.000000   
17                     India Bazaar, The Beaches West     0.000000  0.000000   
18          Kensington Market, Chinatown, Grange Park     0.000000  0.000000   
19                                      Lawrence Park     0.000000  0.000000   
20                           Little Portugal, Trinity     0.023810  0.000000   
21                        Moore Park, Summerhill East     0.000000  0.000000   
22                                 North Toronto West     0.052632  0.000000   
23                             Parkdale, Roncesvalles     0.000000  0.000000   
24                          Regent Park, Harbourfront     0.022222  0.000000   
25                           Richmond, Adelaide, King     0.000000  0.000000   
26                                           Rosedale     0.000000  0.000000   
27                                           Roselawn     0.000000  0.000000   
28                                 Runnymede, Swansea     0.027778  0.000000   
29                                     St. James Town     0.000000  0.000000   
30                        St. James Town, Cabbagetown     0.022727  0.000000   
31                                    Studio District     0.027778  0.000000   
32  Summerhill West, Rathnelly, South Hill, Forest...     0.000000  0.000000   
33                The Annex, North Midtown, Yorkville     0.000000  0.000000   
34                                        The Beaches     0.000000  0.000000   
35                                 The Danforth  East     0.000000  0.000000   
36                       The Danforth West, Riverdale     0.023810  0.000000   
37           Toronto Dominion Centre, Design Exchange     0.000000  0.000000   
38                     University of Toronto, Harbord     0.030303  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000              0.0   
1             0.000000      0.000000        0.000000              0.0   
2             0.066667      0.066667        0.066667              0.2   
3             0.000000      0.000000        0.000000              0.0   
4             0.000000      0.000000        0.000000              0.0   
5             0.000000      0.000000        0.000000              0.0   
6             0.000000      0.000000        0.000000              0.0   
7             0.000000      0.000000        0.000000              0.0   
8             0.000000      0.000000        0.000000              0.0   
9             0.000000      0.000000        0.000000              0.0   
10            0.000000      0.000000 

In [44]:
print(toronto_grouped.shape)

(39, 231)


### Checking Top 5 venues in each neighborhood by looking at frequencies

In [45]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
            venue  freq
0     Coffee Shop  0.05
1          Bakery  0.05
2    Cocktail Bar  0.05
3        Pharmacy  0.03
4  Farmers Market  0.03


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.14
1  Breakfast Spot  0.09
2     Coffee Shop  0.09
3          Bakery  0.05
4       Pet Store  0.05


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                venue  freq
0     Airport Service  0.20
1     Harbor / Marina  0.07
2       Boat or Ferry  0.07
3  Airport Food Court  0.07
4        Airport Gate  0.07


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.18
1      Sandwich Place  0.06
2                Café  0.06
3          Restaurant  0.04
4  Italian Restaurant  0.04


----Christie----
           venue  freq
0  Grocery Store  0.25
1           Café  0.19
2           Park  0.12
3      Nightclub  0.06
4    Candy Store  0.0

### Function that gives top 5 venues in each neighborhood

In [46]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### Turning this information into a dataset

In [49]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park                Bakery   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Coffee Shop          Cocktail Bar        Farmers Market   
1           Coffee Shop        Breakfast Spot                   Gym   
2                 Plane                   Bar           Coffee Shop   
3        Sandwich Place                  Café    Italian Restaurant   
4                  Café                  Park            Restaurant   

  5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0              Pharmacy     Seafood Restaurant            Restaurant   
1          Intersection  Performing Arts Venue             Pet Store   
2   Rental Car Location       Sculpture Garden         Boat or Ferry   
3            Restaurant                    Spa          Burger Joint   
4    Italian Restaurant     Athletics & Sports           Candy Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Beer Bar           Cheese Shop                    Pub  
1         Grocery Store             Nightclub           Climbing Gym  
2       Harbor / Marina        Airport Lounge     Airport Food Court  
3   Japanese Restaurant       Bubble Tea Shop            Salad Place  
4             Nightclub            Baby Store            Coffee Shop

# Clustering Neighborhoods

In [50]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

### Merging datasets

In [51]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
toronto_merged = toronto_data
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() 

Borough               Neighborhood PostalCode   Latitude  \
2   Downtown Toronto  Regent Park, Harbourfront        M5A  43.654260   
9   Downtown Toronto   Garden District, Ryerson        M5B  43.657162   
15  Downtown Toronto             St. James Town        M5C  43.651494   
19      East Toronto                The Beaches        M4E  43.676357   
20  Downtown Toronto                Berczy Park        M5E  43.644771   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
2  -79.360636               1           Coffee Shop                Bakery   
9  -79.378937               1           Coffee Shop        Clothing Store   
15 -79.375418               1           Coffee Shop                  Café   
19 -79.293031               1     Health Food Store                 Trail   
20 -79.373306               1                Bakery           Coffee Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2                    Pub                  Park                  Café   
9    Japanese Restaurant       Bubble Tea Shop                  Café   
15        Cosmetics Shop                 Hotel            Restaurant   
19                   Pub             Wine Shop                 Diner   
20          Cocktail Bar        Farmers Market              Pharmacy   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2         Breakfast Spot               Theater   Distribution Center   
9         Cosmetics Shop  Fast Food Restaurant             Bookstore   
15        Clothing Store          Cocktail Bar                  Park   
19    Falafel Restaurant           Event Space  Ethiopian Restaurant   
20    Seafood Restaurant            Restaurant              Beer Bar   

   9th Most Common Venue       10th Most Common Venue  
2      Electronics Store           Mexican Restaurant  
9       Ramen Restaurant                        Hotel  
15    Seafood Restaurant                     Creperie  
19     Electronics Store  Eastern European Restaurant  
20           Cheese Shop                          Pub

# Cluster Map

There are a lot of similar neighborhoods which is seen in purple. In addition, there are few different ones which are blue,yellow and red. This means that most of the neighborhoods are same in terms of venues. If I want to choose one, I should look at the location of office and venues which I most like specifiaclly.

In [52]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examining Clusters

### Cluster 1

In [54]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
35  The Danforth  East               0           Pizza Place   
61       Lawrence Park               0                  Park   
91            Rosedale               0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
35     Convenience Store                  Park        Discount Store   
61              Bus Line           Swim School             Wine Shop   
91            Playground                 Trail             Wine Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
35        Farmers Market    Falafel Restaurant           Event Space   
61   Distribution Center        Farmers Market    Falafel Restaurant   
91                 Diner    Falafel Restaurant           Event Space   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
35  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant  
61           Event Space  Ethiopian Restaurant            Electronics Store  
91  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant

### Cluster 2

In [55]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
2                            Regent Park, Harbourfront               1   
9                             Garden District, Ryerson               1   
15                                      St. James Town               1   
19                                         The Beaches               1   
20                                         Berczy Park               1   
24                                  Central Bay Street               1   
25                                            Christie               1   
30                            Richmond, Adelaide, King               1   
31                        Dufferin, Dovercourt Village               1   
36   Harbourfront East, Union Station, Toronto Islands               1   
37                            Little Portugal, Trinity               1   
41                        The Danforth West, Riverdale               1   
42            Toronto Dominion Centre, Design Exchange               1   
43        Brockton, Parkdale Village, Exhibition Place               1   
47                      India Bazaar, The Beaches West               1   
48                      Commerce Court, Victoria Hotel               1   
54                                     Studio District               1   
67                                    Davisville North               1   
69                       High Park, The Junction South               1   
73                                  North Toronto West               1   
74                 The Annex, North Midtown, Yorkville               1   
75                              Parkdale, Roncesvalles               1   
79                                          Davisville               1   
80                      University of Toronto, Harbord               1   
81                                  Runnymede, Swansea               1   
84           Kensington Market, Chinatown, Grange Park               1   
86   Summerhill West, Rathnelly, South Hill, Forest...               1   
87   CN Tower, King and Spadina, Railway Lands, Har...               1   
92                                      Enclave of M5E               1   
96                         St. James Town, Cabbagetown               1   
97              First Canadian Place, Underground city               1   
99                                Church and Wellesley               1   
100                                     Enclave of M4L               1   

    1st Most Common Venue  2nd Most Common Venue 3rd Most Common Venue  \
2             Coffee Shop                 Bakery                   Pub   
9             Coffee Shop         Clothing Store   Japanese Restaurant   
15            Coffee Shop                   Café        Cosmetics Shop   
19      Health Food Store                  Trail                   Pub   
20                 Bakery            Coffee Shop          Cocktail Bar   
24            Coffee Shop         Sandwich Place                  Café   
25          Grocery Store                   Café                  Park   
30            Coffee Shop         Clothing Store                  Café   
31                 Bakery               Pharmacy             Wine Shop   
36            Coffee Shop               Aquarium                  Café   
37                    Bar  Vietnamese Restaurant                  Café   
41       Greek Restaurant            Coffee Shop    Italian Restaurant   
42            Coffee Shop                  Hotel                  Café   
43                   Café            Coffee Shop        Breakfast Spot   
47         Sandwich Place                   Park                   Pub   
48            Coffee Shop             Restaurant                  Café   
54            Coffee Shop                   Café                Bakery   
67      Food & Drink Shop           Dance Studio  Gym / Fitness Center   
69                   Park     Mexican Restaurant                  Café   
73         Clothing Store            Coffee S

### Cluster 3

In [56]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
83  Moore Park, Summerhill East               2          Intersection   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
83            Restaurant             Wine Shop        Farmers Market   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
83    Falafel Restaurant           Event Space  Ethiopian Restaurant   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
83     Electronics Store  Eastern European Restaurant    Dumpling Restaurant

### Cluster 4

In [57]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
62     Roselawn               3  Fast Food Restaurant                Garden   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
62             Wine Shop   Distribution Center        Farmers Market   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
62    Falafel Restaurant           Event Space  Ethiopian Restaurant   

   9th Most Common Venue       10th Most Common Venue  
62     Electronics Store  Eastern European Restaurant

### Cluster 5

In [58]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
68  Forest Hill North & West               4                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
68    Mexican Restaurant         Jewelry Store      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
68             Wine Shop   Distribution Center        Farmers Market   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
68    Falafel Restaurant           Event Space   Ethiopian Restaurant